File: src/java/magma/Result.java
package magma;

import java.io.IOException;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public interface Result<T, X> {
    Optional<T> findValue();

    Optional<X> findError();

    <R> Result<Tuple<T, R>, X> and(Supplier<Result<R, X>> other);

    <R> Result<R, X> mapValue(Function<T, R> mapper);

    <R> Result<R, X> flatMapValue(Function<T, Result<R, X>> mapper);

    boolean isOk();

    <R> Result<T, R> mapErr(Function<X, R> mapper);

    <R> R match(Function<T, R> whenOk, Function<X, R> whenErr);
}

File: src/windows/magma/api/result/Result.c
#include "./Result.h"
export interface Result<T, X> {
	Option<X> findError();
	Option<T> findValue();
	R match((arg0 : T) => R whenOk, (arg0 : X) => R whenErr);
	Result<R, X> flatMapValue((arg0 : T) => Result<R, X> mapper);
	Result<R, X> mapValue((arg0 : T) => R mapper);
}
